pkgs/by-name/po/porn-vault/porn-vault.sh

```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+cd $(mktemp -d)
```

Comment:
I think the general approach when apps need such an intermingling of their static assets, config files, and state, is to put it somewhere under the service's `StateDirectory`. Home Assistant does something similar, see https://github.com/NixOS/nixpkgs/blob/689fed12a013f56d4c4d3f612489634267d86529/nixos/modules/services/home-automation/home-assistant.nix#L488

Since `mktemp` things are not cleaned up automatically (AFAIK), this may cause disk space to balloon after a while for people who have a persistent `/tmp`.

In any case, it would be ideal to minimize these file operations as much as possible, but I'm not familiar with PV and so do not know how easy that would be.

Reply:
PV expects to be run on the same workdir as it is stored, so you can’t just do “node $out/…” without cding into it because it will crash.
But we can’t just cd into store, because PV also creates a tmp folder on your current workdir. So if you run it cding into store, it will fail due to the read-only filesystem.

A different approach I used was creating a static /tmp/porn-vault folder and making a symbolic link from $out/share/porn-vault/tmp targeting there. This avoids all the copying (symlinking, actually) stuff, but might interfere if the user were to run more than one PV instance at the same time. Anyhow, I’m not sure how much sense it would make for one to run more than one PV at the same time or if that is even possible (even with the current approach) at all. 

If this is preferred, at installPhase I’d just run the ln command from $out targeting /tmp (is that allowed? A symlink at store targeting something mutable?). On this case, the startup script would be limited into something along the lines of 

```sh
mkdir -p /tmp/porn-vault
cd $out/share/porn-vault
[[ -z "${PV_CONFIG_FOLDER}" ]] || install -Dm111 %izzy%/bin/izzy $PV_CONFIG_FOLDER/izzy
exec %nodejs%/bin/node ./dist/index.js
```

Do note that, tho, the current script doesn’t copy much stuff at all. It makes symbolic links from the mktemp folder. Of course, if done enough without cleaning up then the symlinks could theoretically occupy some disk space, but I’m not sure how realistic that expectation actually is.
The only thing that is copied is the izzy binary, but note that that only happens once! The binary is copied into the user’s config folder, not the tmp folder, so it won’t contribute to the mentioned problem at all. Everything in the mktemp folder is a symbolic link (other than the tmp folder porn-vault makes, but that’s usually empty when porn-vault is powered off (it’s just used during transcoding I think)), nothing is actually copied there at every run.

Still, if symbolic linking $out/share/porn-vault/tmp to /tmp/porn-vault is allowed I think I’d still prefer that approach. I can confirm it works because it was the first approach I tried.

Reply:
If PV requires it, then I think the current approach makes way more sense than trying to link something under `$out` to `/tmp/...`. So my only doubt is whether this "link farm" should be a `mktemp`-d directory, or something like the unit's `StateDirectory`. I'll defer to someone more experienced on this question. As mentioned, for example Home Assistant uses `/var/lib/hass` to similarly link stuff like configuration YAML-s, although the application itself is not linked in that case methinks.

Reply:
I could do that, but then would standalone PV running still work?

Reply:
Just thought about a different approach perhaps. If /tmp is actually not automatically cleared, and I can see why that’s a concern, we could name the tmp folders as "/tmp/porn-vault-$(mktemp -u)". Then every time the service starts it could clean up previous runs stuff, with "rm -rf /tmp/porn-vault-*" (pretty sure that will not try to follow the symlinks), before creating the new folder.

Would that be an acceptable approach? This way we’d still allow for PV to be ran separately from the module, but it’d clean up stuff.

Also, note that on the service I’m using systemd hardening with PrivateTmp = true;. Would the service still have access to its own /tmp of previous runs? Otherwise the above approach will probably be unnecessary, assuming that systemd deleted the previous PrivateTmp.

Reply:
> PrivateTmp=
> Takes a boolean argument. If true, sets up a new file system namespace for the executed processes and mounts private /tmp/ and /var/tmp/ directories inside it that are not shared by processes outside of the namespace. This is useful to secure access to temporary files of the process, but makes sharing between processes via /tmp/ or /var/tmp/ impossible. **If true, all temporary files created by a service in these directories will be removed after the service is stopped. **

[(source)](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#)

So yeah, that’s actually not needed

(except perhaps when running separately from the service. On that case, maybe do the tmp folder name plus rm thing?)